skops/cli/_convert.py

```diff
@@ -0,0 +1,114 @@
+from __future__ import annotations
+
+import argparse
+import logging
+import os
+import pathlib
+import pickle as pkl
+from typing import Optional
+
+from skops.io import dumps, get_untrusted_types
+
+
+def _convert_file(input_file: os.PathLike, output_file: pathlib.Path):
+    """
+    Function that is called by ``skops convert`` entrypoint.
+
+    Loads a pickle model from the input path, converts to skops format, and saves to
+    output file.
+
+    Parameters
+    ----------
+    input_file : os.PathLike
+        Path of input .pkl model to load.
+
+    output_file : os.PathLike
+        Path to save .skops model to.
+
+    """
+    model_name = pathlib.Path(input_file).stem
+
+    logging.info(f"Converting {model_name}")
+
+    with open(input_file, "rb") as f:
+        obj = pkl.load(f)
+    skops_dump = dumps(obj)
+
+    untrusted_types = get_untrusted_types(data=skops_dump)
+
+    if not untrusted_types:
+        logging.debug(f"No unsafe types found in {model_name}.")
+    else:
+        untrusted_str = "\n".join(untrusted_types)
+
+        logging.warning(
+            "Unknown Types Detected!\n"
+            f"While converting {model_name}, "
+            "the following unsafe types were found: \n"
+            f"{untrusted_str}\n\n"
+            f"When loading{output_file}, add ``--trusted`` to the skops.load call."
+        )
+
+    with open(output_file, "wb") as out_file:
+        logging.info(f"Writing to {output_file}")
+        out_file.write(skops_dump)
+
+
+def main_convert(
+    command_line_args: Optional[list[str]] = None,
+    parent: Optional[argparse.ArgumentParser] = None,
+):
+    parents = [parent] if parent else []
+
+    parser = argparse.ArgumentParser(
+        description="Convert input Pickle files to .skops files", parents=parents
+    )
+
+    parser.add_argument("inputs", nargs="+", help="Input files to convert.")
+    parser.add_argument(
+        "-o",
+        "--output-files",
+        nargs="+",
+        help=(
+            "Specify output file names for the converted skops files."
+            "If not provided, will default to using the same name as the input file, "
+            "and saving to the current working directory."
+        ),
+        default=None,
+    )
+    parser.add_argument(
+        "-d",
+        "--debug",
+        help="Enable debug logging.",
+        action="store_const",
+        dest="loglevel",
+        const=logging.DEBUG,
+        default=logging.WARNING,
+    )
+    parser.add_argument(
```

Comment:
Where is `verbose` used? Do we need `verbose` and `debug`?

Reply:
The normal pattern I've seen used before with varying log levels is to tie `--verbose` to `info` level logging, and `--debug` to debug level logging.

So in the current code, with the `-v` flag, the info, debug and above level statements will be written out, and with `-d`, only debug and above level logs are written out.

(The info level logging is really fine grained: it logs when a model starts, and finishes being converted. Adjusting the log level here also lets any logging used elsewhere in our codebase get handled correctly.)

Reply:
Ah I see how it's used now.

> The normal pattern I've seen used before with varying log levels is to tie --verbose to info level logging, and --debug to debug level logging.

Which libraries do that? Could you please give me a pointer?

> So in the current code, with the -v flag, the info, debug and above level statements will be written out, and with -d, only debug and above level logs are written out.

INFO level is 20, DEBUG level is 10, so when setting `-v`, debug is not shown, and when setting `-d`, debug and info are shown. 

Furthermore, setting both `-v` and `-d` sets the log level to INFO, which as a user, I would find confusing. That's why I would prefer having a single argument be in control of the granularity of the logging. That could be a `loglevel` argument, where users can pass `INFO` or `DEBUG` or an int. Or it could be an incremental `verbose` argument, where  users can pass `-v`, `-vv`, etc.

> Adjusting the log level here also lets any logging used elsewhere in our codebase get handled correctly

I don't understand this point.

Reply:
> INFO level is 20, DEBUG level is 10, so when setting -v, debug is not shown, and when setting -d, debug and info are shown.

I always get that mixed up, sorry. I can shift the info and debug level logs.

The usual pattern I've seen used is described [here in the Ubuntu community docs](https://discourse.ubuntu.com/t/cli-verbosity-levels/26973). Here you can see varied level outputs from quiet -> trace, but essentially:

* Verbose logs should include more detailed descriptions
* Debug logs should describe internal steps as they happen to help in debugging

And verbose typically uses `-v` or `--verbose` as flags, whereas debug uses `-d` or `--debug`.

However, this is for Linux CLIs, and from what I can tell, many python packages tend to prefer the incremental `verbose` pattern you described.

> Adjusting the log level here also lets any logging used elsewhere in our codebase get handled correctly

Basically I just meant we can add more logging in the future without having to manually add logic to handle it. 

With log levels, we can just add `logging.debug(...)` in the skops.io module and given the right log level, that will be output in the CLI without needing any new logic.

Reply:
> The usual pattern I've seen used is described [here in the Ubuntu community docs](https://discourse.ubuntu.com/t/cli-verbosity-levels/26973). Here you can see varied level outputs from quiet -> trace, but essentially:

I found this to be unclear. In the table, it appears that the different options describe an increasing *scale* of verbosity. In that case, I would argue for a single option that controls verbosity, since we're talking about changing a single value.

On the other hand, in the text it says:

> In verbose mode, you want to talk to the audiences in a more descriptive and thorough description about what happens in transactions or a series of events. This could be a conversation log between a user/client and the machine. [...]
> In debug mode, the mental model is to think of 2 developers talking to one another trying to understand where things go wrong. [...]

This reads like different *types* of output, one designed for users, one for devs, not something on a scale. If this was the goal, I can see having two different options, `verbose` for users and `debug` for devs, which can be turned on/off independently and don't overlap. But this would *not* map onto `logging.INFO` and `logging.DEBUG`, since those are on a scale.

Can we maybe not simplify the CLI for now by only providing a default mode, which is to not print anything, except if an error occurs, and a `verbose` mode, which prints info level stuff (but not debug)?

> Basically I just meant we can add more logging in the future without having to manually add logic to handle it.

Got it.

Reply:
> Can we maybe not simplify the CLI for now by only providing a default mode, which is to not print anything, except if an error occurs, and a verbose mode, which prints info level stuff (but not debug)?

Happy to do this. 

> This reads like different types of output, one designed for users, one for devs, not something on a scale. If this was the goal, I can see having two different options, verbose for users and debug for devs, which can be turned on/off independently and don't overlap. But this would not map onto logging.INFO and logging.DEBUG, since those are on a scale.

I think *broadly* they do exist on a scale (at least in Ubuntu-land), where "verbose" means "log things up to INFO level" and "debug" means "log things up to DEBUG level", as well as the changes that can be seen in the table on that page of what get logged where. The way I read that description in the text was more about what logs at INFO and DEBUG level should look like.

I did more digging, however, and have seen that in [Android logging](https://developer.android.com/studio/debug/am-logcat), "verbose" means "write out ALL level logs", so am starting to think the "canonical" way to tell a CLI what log level just depends on who you ask.

Happy to go to a much simpler and just go with number of "v"s to adjust log level.